src/handlers/web3.ts

```diff
@@ -0,0 +1,683 @@
+import {
+  ARBITRUM_ETH_ADDRESS,
+  ETH_ADDRESS,
+  ethUnits,
+  MATIC_POLYGON_ADDRESS,
+  OPTIMISM_ETH_ADDRESS,
+  smartContractMethods,
+} from '@rainbow-me/references';
+import {
+  addBuffer,
+  convertAmountToRawAmount,
+  convertStringToHex,
+  fraction,
+  greaterThan,
+  handleSignificantDecimals,
+  multiply,
+} from '@rainbow-me/utilities';
+import { ethereumUtils } from '@rainbow-me/utils';
+import logger from 'logger';
+
+export const networkProviders: { [network in Network]?: JsonRpcProvider } = {};
+
+const rpcEndpoints: { [network in Network]?: string } = {};
+
+/**
+ * The format of transaction details returned by functions such as `getTxDetails`.
+ */
+interface TransactionDetailsReturned {
+  data?: TransactionRequest['data'];
+  to?: TransactionRequest['to'];
+  value?: TransactionRequest['value'];
+  from?: TransactionRequest['from'];
+  gasLimit?: string;
+  gasPrice?: string;
+}
+
+/**
+ * A type with the same keys as `NewTransaction` but without nullable types.
+ * This is useful for functions that assume that certain fields are not set
+ * to null on a `NewTransaction`.
+ */
+type NewTransactionNonNullable = {
+  [key in keyof NewTransaction]: NonNullable<NewTransaction[key]>;
+};
+
+/**
+ * @desc Configures `rpcEndpoints` based on a given `RainbowConfig`.
+ * @param config The `RainbowConfig` to use.
+ */
+export const setRpcEndpoints = (config: RainbowConfig) => {
+  rpcEndpoints[Network.mainnet] = config.ethereum_mainnet_rpc;
+  rpcEndpoints[Network.ropsten] = config.ethereum_ropsten_rpc;
+  rpcEndpoints[Network.kovan] = config.ethereum_kovan_rpc;
+  rpcEndpoints[Network.rinkeby] = config.ethereum_rinkeby_rpc;
+  rpcEndpoints[Network.optimism] = config.optimism_mainnet_rpc;
+  rpcEndpoints[Network.arbitrum] = config.arbitrum_mainnet_rpc;
+  rpcEndpoints[Network.polygon] = config.polygon_mainnet_rpc;
+};
+
+/**
+ * @desc web3 http instance
+ */
+export let web3Provider: JsonRpcProvider = (null as unknown) as JsonRpcProvider;
+
+/**
+ * @desc Checks whether or not a `Network | string` union type should be
+ * treated as a `Network` based on its prefix, as opposed to a `string` type.
+ * @param network The network to check.
+ * @returns A type predicate of `network is Network`.
+ */
+const isNetworkEnum = (network: Network | string): network is Network => {
+  return !network.startsWith('http://');
+};
+
+/**
+ * @desc set a different web3 provider
+ * @param {String} network
+ */
+export const web3SetHttpProvider = async (network: Network | string) => {
+  if (!isNetworkEnum(network)) {
+    web3Provider = new JsonRpcProvider(network, 'any');
+    // override mainnet for hardhat
+    networkProviders[Network.mainnet] = web3Provider;
+  } else {
+    web3Provider = new JsonRpcProvider(rpcEndpoints[network]);
+  }
+  return web3Provider.ready;
+};
+
+/**
+ * @desc returns true if the given network is a Layer 2
+ * @param network The network to check.
+ */
+export const isL2Network = (network: Network | string) => {
```

Comment:
https://github.com/rainbow-me/rainbow/pull/2511/files#diff-1df2f9097c6797d911776c84fe906260b0e621ac347c2e81781b1f22b8385e55R84

is this the only reason for `| string`? Can we made this `null` instead?

Reply:
I think that ultimately we should not accept `string` in places like `web3SetHttpProvider`, `isNetworkEnum`, `isL2Network` etc.

Reply:
hey I can't seem to see the file diff correctly so I'm not sure if we're referring to the same thing - the string is there for the hardhat option for debug and testing . ideally we would want to have hardhat as a separate network, but unfortunately this causes other problems in our codebase as of now to ensure that everything else is still considered "mainnet" and not actually separately hardhat